concourse/client/util.py

```diff
@@ -122,3 +128,73 @@ def jobs_not_triggered(
             for pr_resource_version in pr_resource_versions:
                 if not wait_for_job_to_be_triggered(job, pr_resource_version, concourse_api):
                     yield (job, pr_resource, pr_resource_version)
+
+
+def process_not_triggered_job(
+    job: Job,
+    resource: PipelineConfigResource,
+    resource_version: ResourceVersion,
+    concourse_api
+):
+    cc_resource = concourse_api.resource(
+        pipeline_name=resource.pipeline_name(),
+        resource_name=resource.name,
+    )
+
+    if not cc_resource.is_pinned():
+        pin_resource_trigger_job_unpin(
+            job=job,
+            resource=cc_resource,
+            resource_version=resource_version,
+            concourse_api=concourse_api,
+        )
+    else:
+        # to be implemeted
+        pass
+
+
+def pin_resource_trigger_job_unpin(
```

Comment:
I think this name is confusing. I'd suggest a more general name (e.g. `handle_resource_pinning` or maybe better `ensure_job_run` or similar).

Reply:
done

Reply:
+1 for a clearer name. as already commented, please consider turning this to a context-mgr. this is i think very much what context-mgrs are intended for. also, the naming would be easier.

consider this code snippet:
```
while True:
  try:
    with pin_resource(...args):
     trigger_job...
     break # finally, we succeeded
  except PinningFailed:
   # too bad, let's retry..
  sleep(....)
  if no_retries_left:
    bail out / fail / error
```